src/app/shared/services/api/test-utils.ts

```diff
@@ -0,0 +1,30 @@
+export function buildUrlWithParams(
+  baseUrl: string,
+  queryParams: Record<string, any>,
+): string {
+  const params = new URLSearchParams();
+
+  const buildParam = (key: string, value: any) => {
+    if (Array.isArray(value)) {
+      value.forEach((v, index) => {
+        if (typeof v === 'object' && v !== null) {
+          Object.entries(v).forEach(([k, val]) =>
+            params.append(`${key}[${index}][${k}]`, String(val)),
+          );
+        } else {
+          params.append(`${key}[]`, String(v));
+        }
+      });
+    } else if (typeof value === 'object' && value !== null) {
+      Object.entries(value).forEach(([k, val]) =>
+        params.append(`${key}[${k}]`, String(val)),
+      );
+    } else {
+      params.append(key, String(value));
+    }
+  };
+
+  Object.entries(queryParams).forEach(([key, value]) => buildParam(key, value));
```

Comment:
I think this function would be more readable if you modified `buildParam` to have the signature `(params: URLSearchParams, [key: string, value: any]): URLSearchParams`, and then used a `reduce` operation here. It's easier to grasp what a function does when it doesn't have side effects, and this would also allow you to define `buildParam` outside of `buildUrlWithParams`.